### PR TITLE
[Serving] Add flow topology support

### PR DESCRIPTION
**Pull request by yaronha**
* add support for flows/DAG in serving runtime
* enhance the graph, support streams, multi-function, ..
* work with Storey as async runtime engine

### CODE REVIEW

**Review by Hedingber — CHANGES_REQUESTED** [review]
Looks good
The main thing I have to say is that without very high test coverage no one but you will be able to touch that code.
This PR does add very nice coverage, It was hard for me to follow and verify exactly what covered and what's not, if there is something you know that is not, please add
Also I highly suggest system tests to the whole serving thing

File: mlrun/runtimes/serving.py [hunk 1]
```diff
@@ -115,22 +125,37 @@ def __init__(
...
         return self._graph
 
     @graph.setter
     def graph(self, graph):
-        self._graph = self._verify_dict(graph, "graph", ServingRouterState)
+        graph_root_setter(self, graph)
+
+    @property
+    def function_refs(self) -> List[FunctionReference]:
+        """"function references, list of optional child function refs"""
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:153]: done

File: mlrun/runtimes/serving.py [hunk 2]
```diff
@@ -145,19 +170,43 @@ def spec(self, spec):
...
+            graph = fn.set_topology("flow", engine="async")
+            graph.to("MyClass").to(name="to_json", handler="json.dumps").respond()
+
+        topology can be:
+          router - root router + multiple child route states/models
+                   route is usually determined by the path (route key/name)
+                   can specify special router class and router arguments
+
+          flow   - workflow (DAG) with a chain of states
+                   flow support "sync" and "async" engines, branches are note allowed in sync mode
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:187]: done

File: mlrun/runtimes/serving.py [hunk 3]
```diff
@@ -145,19 +170,43 @@ def spec(self, spec):
...
+                   when using async mode .respond() can indicate the state which will
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:188]: explained better

File: mlrun/runtimes/serving.py [hunk 4]
```diff
@@ -145,19 +170,43 @@ def spec(self, spec):
...
-            raise NotImplementedError("currently only supporting router topology")
-        self.spec.graph = ServingRouterState(
-            class_name=class_name, class_args=class_args
-        )
+        if topology == StateKinds.router:
+            self.spec.graph = RouterState(class_name=class_name, class_args=class_args)
+        elif topology == StateKinds.flow:
+            self.spec.graph = RootFlowState(engine=engine)
+        else:
+            raise ValueError(f"unsupported topology {topology}, use 'router' or 'flow'")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:208]: done

File: mlrun/runtimes/serving.py [hunk 5]
```diff
@@ -205,58 +264,114 @@ def add_model(
...
-            route = new_model_endpoint(class_name, model_path, handler, **class_args)
-        self.spec.graph.add_route(key, route)
+            state = new_model_endpoint(class_name, model_path, handler, **class_args)
 
-    def remove_models(self, keys: list):
-        """remove one, multiple, or all models from the spec (blank list for all)"""
+        return graph.add_route(key, state)
+
+    def add_child_function(
+        self, name, url=None, image=None, requirements=None, kind=None
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:275]: its much easier for user to to do fn.add_..() vs import/create FuncRef.. will add doc string to it to make it even simpler

File: mlrun/runtimes/serving.py [hunk 6]
```diff
@@ -205,58 +264,114 @@ def add_model(
...
+            url, image, requirements=requirements, kind=kind or "serving"
+        )
+        self._spec.function_refs.update(function_reference, name)
+        func = function_reference.to_function(self.kind)
+        func.set_env("SERVING_CURRENT_FUNCTION", function_reference.name)
+        return func
+
+    def _add_ref_triggers(self):
+        """add stream trigger to downstream child functions"""
+        for function, stream in self.spec.graph.get_queue_links().items():
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:288]: done

File: mlrun/runtimes/serving.py [hunk 7]
```diff
@@ -205,58 +264,114 @@ def add_model(
...
+                group = stream.options.get("group", "serving")
+
+                child_function = self._spec.function_refs[function]
+                child_function.function_object().add_stream_trigger(
+                    stream.path, group=group, shards=stream.shards
+                )
+
+    def _deploy_function_refs(self):
+        """set metadata and deploy child functions"""
+        for function in self._spec.function_refs.values():
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:301]: done

File: mlrun/runtimes/serving.py [hunk 8]
```diff
@@ -205,58 +264,114 @@ def add_model(
...
+            self._add_ref_triggers()
+            self._deploy_function_refs()
+            logger.info(f"deploy root function {self.metadata.name} ...")
+        return super().deploy(dashboard, project, tag, verbose=verbose)
 
     def _get_runtime_env(self):
-        # we currently support a minimal topology of one router + multiple child routes/models
-        # in the future we will extend the support to a full graph, the spec is already built accordingly
+
+        functions = {f.name: f.uri(self) for f in self.spec.function_refs}
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/runtimes/serving.py:346]: done

File: mlrun/serving/server.py [hunk 9]
```diff
@@ -50,119 +57,148 @@ def __init__(
...
         self.verbose = verbose
         self.load_mode = load_mode or "sync"
         self.version = version or "v2"
         self.context = None
-        self._namespace = None
+        self._current_function = None
+        self.functions = functions or {}
+        self.graph_initializer = graph_initializer
+        self.error_stream = error_stream
+        self._error_stream = None
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/serving/server.py:76]: yes, error stream is a stream that collect all the graph exceptions (requested by customers), graph_initializer is used by the feature store to enrich/init the context w app specific stuff (like nuclio init_context), ..

File: mlrun/serving/states.py [hunk 10]
```diff
@@ -347,8 +1020,187 @@ def get_class(class_name, namespace):
     return class_object
 
 
+def get_function(function, namespace):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/serving/states.py:1023]: i will move those 3 functions (get func/module/class) to helpers (near create_class() ..)

File: mlrun/serving/states.py [hunk 11]
```diff
@@ -347,8 +1020,187 @@ def get_class(class_name, namespace):
...
 }
+
+
+def get_current_function(context):
+    if context and hasattr(context, "current_function"):
+        return context.current_function or ""
+    return ""
+
+
+def _add_gviz_router(g, state, source=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/serving/states.py:1058]: changed to `graphviz`, g is the graphviz Digraph object, its a common practice to name it `g`

File: mlrun/serving/states.py [hunk 12]
```diff
@@ -347,8 +1020,187 @@ def get_class(class_name, namespace):
...
+def _generate_graphviz(
+    state, renderer, filename=None, format=None, source=None, targets=None, **kw,
+):
+    try:
+        from graphviz import Digraph
+    except ImportError:
+        raise ImportError(
+            'graphviz is not installed, run "pip install graphviz" first!'
+        )
+    g = Digraph("mlrun-flow", format="jpg")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/serving/states.py:1115]: changed to `graph`

File: tests/serving/test_async_flow.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+def has_storey():
+    try:
+        import storey  # noqa
+    except ImportError:
+        return False
+    return True
+
+
+@pytest.mark.skipif(not has_storey(), reason="storey not installed")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:14]: done

File: tests/serving/test_async_flow.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+def test_handler():
+    fn = mlrun.new_function("tests", kind="serving")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:16]: done

File: tests/serving/test_async_flow.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    graph = fn.set_topology("flow", engine="async")
+    graph.add_step(name="s1", handler="(event + 1)")
+    graph.add_step(name="s2", handler="json.dumps", after="$prev").respond()
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:19]: the test is to test handlers (lambda and dumps), the json.dumps turns the `6` to a str (`"6"`)

File: tests/serving/test_async_flow.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    # flow.plot("async.png")
+
+    server = fn.to_mock_server()
+    server.context.visits = {}
+    print("\nAsync Flow:\n", flow.to_yaml())
+    resp = server.test(body=[])
+    print(resp)
+
+    server.wait_for_completion()
+    assert resp == ["s1", "s2", "s5"], "flow result is incorrect"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:48]: s5 is the responder, so only the data collected in the s5 branch are returned in resp. vs visited (all steps were visited)

File: tests/serving/test_async_flow.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+@pytest.mark.skipif(not has_storey(), reason="storey not installed")
+def test_async_nested():
+    fn = mlrun.new_function("tests", kind="serving")
+    graph = fn.set_topology("flow", engine="async")
+    graph.add_step(name="s1", class_name="Echo")
+    graph.add_step(name="s2", handler="multiply_input", after="s1")
+    graph.add_step(name="s3", class_name="Echo", after="s2")
+
+    router = graph.add_step("*", name="ensemble", after="s2")
+    router.add_route("m1", class_name="ModelClass", model_path=".", z=100)
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:67]: done

File: tests/serving/test_async_flow.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+@pytest.mark.skipif(not has_storey(), reason="storey not installed")
+def test_async_nested():
+    fn = mlrun.new_function("tests", kind="serving")
+    graph = fn.set_topology("flow", engine="async")
+    graph.add_step(name="s1", class_name="Echo")
+    graph.add_step(name="s2", handler="multiply_input", after="s1")
+    graph.add_step(name="s3", class_name="Echo", after="s2")
+
+    router = graph.add_step("*", name="ensemble", after="s2")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:66]: do think it really change (router vs router step), but changed it

File: tests/serving/test_async_flow.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+    router.add_route("m3:v1", class_name="ModelClass", model_path=".", z=300)
+
+    graph.add_step(name="final", class_name="Echo", after="ensemble").respond()
+
+    print(graph.to_yaml())
+    server = fn.to_mock_server()
+    # graph.plot("nested.png")
+    resp = server.test("/v2/models/m2/infer", body={"inputs": [5]})
+    server.wait_for_completion()
+    assert resp["outputs"] == 2000, f"wrong health response {resp}"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_async_flow.py:78]: fixed

File: tests/serving/test_serving.py [hunk 20]
```diff
@@ -231,7 +224,22 @@ def test_v2_health():
...
-    host = create_mock_server()
-    host.add_model("my", class_name=ModelTestingClass, model_path="", z=100)
-    print(host.test("my/infer", testdata))
+    host = create_graph_server(graph=RouterState())
+    host.graph.add_route("my", class_name=ModelTestingClass, model_path="", z=100)
+    host.init(None, globals())
     print(host.to_yaml())
+    resp = host.test("/v2/models/my/infer", testdata)
+    print(resp)
+    assert resp["outputs"] == 500, f"wrong health response {resp}"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_serving.py:233]: fixed

File: tests/serving/test_serving.py [hunk 21]
```diff
@@ -231,7 +224,22 @@ def test_v2_health():
...
+def test_function():
+    fn = mlrun.new_function("tests", kind="serving")
+    graph = fn.set_topology("router")
+    fn.add_model("my", class_name="ModelTestingClass", model_path=".", z=100)
+
+    server = fn.to_mock_server()
+    # graph.plot("router.png")
+    print("\nFlow:\n", graph.to_yaml())
+    resp = server.test("/v2/models/my/infer", testdata)
+    assert resp["outputs"] == 500, f"wrong health response {resp}"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_serving.py:245]: fixed

File: tests/serving/test_flow.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def test_handler():
+    fn = mlrun.new_function("tests", kind="serving")
+    graph = fn.set_topology("flow", engine="sync")
+    graph.to(name="s1", handler="(event + 1)").to(name="s2", handler="json.dumps")
+
+    server = fn.to_mock_server()
+    resp = server.test(body=5)
+    assert resp == "6", f"got unexpected result {resp}"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply tests/serving/test_flow.py:48]: done

File: tests/serving/test_async_flow.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Hedingber) [reply tests/serving/test_async_flow.py]: This is not what I meant, I meant to replace the `*` (magic value) with something meaningful (like we did `>` to `$queue`)

**Review by Hedingber — CHANGES_REQUESTED** [review]
* Change graphviz requirement to `~=0.16.0`
* you changed `ValueError` only in 2 places, but you have many other places in the code you're adding it, replace to `mlrun.errors...`
* you wanted to rename `self._error_stream`  no ?
* you change `g` to `graph` in `_add_graphviz_flow` but not in `_add_graphviz_router`
* I think that just like you changed `>` to `$queue` you should change `*` to something more meaningful
* If you've left the `flow.plot` everywhere, get it out of comment, change the path to be under `from tests.conftest import results` and add a comment above it "plotting the graph to help debugging"
* `test_on_error` seems duplicated, can be parameterized as well, also I would move the tests that are parameterized to a different file

**Comment by yaronha** [comment]
@Hedingber 
* i can lock `graphviz` but u can see they have a ver every month 
* re the `*` for router its different, its a prefix (e.g. `*Ensemble`) so cant change it to a name, people can always place the class as first arg instead

**Pull request merged**